lib/src/migration_visitor.dart

```diff
@@ -28,74 +26,84 @@ import 'utils.dart';
 /// new instance of itself (using [newInstance]) each time it encounters an
-/// @import or @use rule.
+/// `@import` or `@use` rule.
 abstract class MigrationVisitor extends RecursiveAstVisitor {
-  /// The migrator running on this stylesheet.
-  Migrator get migrator;
-
-  /// The canonical path of the stylesheet being migrated.
-  String get path;
-
   /// The stylesheet being migrated.
-  Stylesheet stylesheet;
+  final Stylesheet stylesheet;
+
+  /// A mapping from URLs to migrated contents for stylesheets already migrated.
+  final Map<Uri, String> migrated;
 
-  /// The syntax this stylesheet uses.
-  Syntax syntax;
+  /// True if dependencies should be migrated as well.
+  final bool migrateDependencies;
 
   /// The patches to be applied to the stylesheet being migrated.
-  final List<Patch> patches = [];
+  UnmodifiableListView<Patch> get patches => UnmodifiableListView(_patches);
+  final List<Patch> _patches = [];
 
-  /// Returns a new instance of this MigrationVisitor with the same migrator
-  /// and [newPath].
-  MigrationVisitor newInstance(String newPath);
+  /// Constructs a new migration visitor, parsing the stylesheet at [url].
+  MigrationVisitor(Uri url, this.migrateDependencies,
+      {Map<Uri, String> migrated})
+      : stylesheet = parseStylesheet(url),
+        this.migrated = migrated ?? {};
+
+  /// Returns a new instance of this MigrationVisitor with the same [migrated]
+  /// and [migrateDependencies] and the new [url].
+  @protected
+  MigrationVisitor newInstance(Uri url);
+
+  /// Adds a new patch that should be applied as part of this migration.
+  @protected
+  void addPatch(Patch patch) {
+    _patches.add(patch);
+  }
 
-  /// Runs the migrator and stores the migrated contents in `migrator.migrated`.
-  void run() {
-    var contents = File(path).readAsStringSync();
-    syntax = Syntax.forPath(path);
-    stylesheet = Stylesheet.parse(contents, syntax, url: path);
+  /// Runs the migrator and returns the map of migrated contents.
+  Map<Uri, String> run() {
     visitStylesheet(stylesheet);
     var results = getMigratedContents();
     if (results != null) {
-      migrator.migrated[path] = results;
+      migrated[stylesheet.span.file.url] = results;
     }
+    return migrated;
   }
 
   /// Returns the migrated contents of this file, or null if the file does not
   /// change.
   ///
   /// This will be called by [run] and the results will be stored in
   /// `migrator.migrated`.
+  @protected
   String getMigratedContents() => patches.isNotEmpty
       ? Patch.applyAll(patches.first.selection.file, patches)
       : null;
 
-  /// Returns the canonical path of [url] when resolved relative to the current
-  /// path.
-  String resolveImportUrl(String url) =>
-      canonicalizePath(p.join(p.dirname(path), url));
+  /// Resolves [relativeUrl] relative to the current stylesheet's URL.
+  @protected
+  Uri resolveRelativeUrl(Uri relativeUrl) =>
+      stylesheet.span.file.url.resolveUri(relativeUrl);
```

Comment:
I prefer referring to a span's URL using [`SourceSpan.sourceUrl`](https://pub.dartlang.org/documentation/source_span/latest/source_span/SourceSpan/sourceUrl.html), since it works for non-`FileSpan`s as well. I think this applies to a couple locations.